src/elements/PolygonAperture.H

```diff
@@ -0,0 +1,296 @@
+        inline std::map<int, amrex::Gpu::DeviceVector<amrex::ParticleReal>> d_vertices_x = {};
+        //! device: skew multipole coefficients of the magnetic field
+        inline std::map<int, amrex::Gpu::DeviceVector<amrex::ParticleReal>> d_vertices_y = {};
+
+    } // namespace PolygonApertureData
+
+    struct PolygonAperture
+    : public mixin::Named,
+      public mixin::BeamOptic<PolygonAperture>,
+      public mixin::LinearTransport<PolygonAperture>,
+      public mixin::Thin,
+      public mixin::Alignment,
+      public mixin::NoFinalize,
+      public amrex::simd::Vectorized<amrex::simd::native_simd_size_particlereal>
+    {
+        static constexpr auto type = "PolygonAperture";
+        using PType = ImpactXParticleContainer::ParticleType;
+
+        enum Action
+        {
+            transmit,
+            absorb
+        };
+
+        static std::string
+        action_name (Action const & action);
+
+        /** A thin collimator element that applies a transverse aperture boundary defined
+         *  by points of a polygon vertices.
+         *  Particles outside the boundary are considered lost.
+         *
+         * @param action specify action of domain (transmit/absorb)
+         * @param vertices_x array of coordinates of the horizontal vertex positions (m)
+         * @param vertices_y array of coordinates of the vertical vertex positions (m)
+         * @param min_radius2 a minimum radius**2 below which all particles are transmitted
+         * @param repeat_x horizontal period for repeated masking, optional (m)
+         * @param repeat_y vertical period for repeated masking, optional (m)
+         * @param shift_odd_x for hexagonal/triangular mask patterns: horizontal shift of every 2nd (odd) vertical period by repeat_x / 2. Use alignment offsets dx,dy to move whole mask as needed.
+         * @param dx horizontal translation error in m
+         * @param dy vertical translation error in m
+         * @param rotation_degree rotation error in the transverse plane [degrees]
+         * @param name a user defined and not necessarily unique name of the element
+         */
+        PolygonAperture (
+            std::vector< amrex::ParticleReal> vertices_xx,
+            std::vector< amrex::ParticleReal> vertices_yy,
+            amrex::ParticleReal min_radius2x,
+            amrex::ParticleReal repeat_x,
+            amrex::ParticleReal repeat_y,
+            bool shift_odd_x,
+            Action action,
+            amrex::ParticleReal dx = 0,
+            amrex::ParticleReal dy = 0,
+            amrex::ParticleReal rotation_degree = 0,
+            std::optional<std::string> name = std::nullopt
+        )
+    //
+        : Named(std::move(name)),
+          Alignment(dx, dy, rotation_degree),
+          m_action(action), m_repeat_x(repeat_x), m_repeat_y(repeat_y), m_shift_odd_x(shift_odd_x),
+      vertices_x(vertices_xx), vertices_y(vertices_yy), min_radius2(min_radius2x)
```

Comment:
There is a small subtlety missing here: You need to initialize the new object's `m_id` for book keeping, like this:
```suggestion
          m_action(action), m_repeat_x(repeat_x), m_repeat_y(repeat_y), m_shift_odd_x(shift_odd_x),
          m_id(PolygonApertureData::next_id),
          vertices_x(vertices_xx), vertices_y(vertices_yy), min_radius2(min_radius2x)
```
Example: https://github.com/BLAST-ImpactX/impactx/blob/25.10/src/elements/SoftSol.H#L174

We essentially use `m_id` to reference global variables outside of this class, in which we store _dynamic_ GPU data (i.e., vectors/arrays of runtime-determined length) of the instances `m_id`. This is primarily because we want to copy the whole of `PolygonAperture` to GPU in the current design, and the dynamic data will only be referenced by pointer for that purpose.

I am aware that this detail is a bit nifty right now, and I am not proud of it ;-) We can refactor the design pattern at some point, but it will likely become more verbose.